Add derivation rules doc (#164)
dwango/fialyzer: docs/derivation-rules.md

This file shows derivation rules used in fialyzer.

Our derivation rules are almost same as [the original paper](https://it.uu.se/research/group/hipe/papers/succ_types.pdf)'s one, but extended by remote call, local call, list, etc.

```
A : context (mapping of variable to type)
τ : type
x : variable
e : term
fun f/a : local function (f: function name, a: arity)
fun m:f/a : remote function (m: module name, f: function name, a: arity)
```

```
-------------------------- [VAR]
A ∪ {x → τ} ⊢ x : τ, ∅

A ⊢ e1 : τ1, C1 ... en : τn, Cn
------------------------------------------------------ [STRUCT]
A ⊢ {e1, ... , en} : {τ1, ... , τn}, C1 ∧ ... ∧ Cn

> A ⊢ e1 : τ1, C1         A ∪ {x → τ1} ⊢ e : τ2, C2

Author: ```suggestion
A ⊢ e1 : τ1, C1         A ∪ {x → τ1} ⊢ e2 : τ2, C2
```

Author: This is an errata of the original paper.

---------------------------------------------------- [LET]
A ⊢ let x = e1 in e2 : τ2, C1 ∧ C2

A ∪ {fun xi/ai → τi} ⊢ f1 : τ'1, C1 ... fn : τ'n, Cn        e : τ, C
---------------------------------------------------------------------------------------------- [LETREC]
A ⊢ letrec x1 = f1, ... , xn = fn in e : τ, C1 ∧ ... Cn ∧ C ∧ (τ'1 = τ1) ∧ ... ∧ (τ'n = τn)
  where ai = length (args fi)

A ∪ {x1 → τ1, ... , xn → τn} ⊢ e : τe, C
---------------------------------------------------------------------- [ABS]
A ⊢ fun(x1, ... , xn) → e : (τ1, ... , τn) → τe, C

A ⊢ e1 : τ1, C1 ... en : τn, Cn
-------------------------------------------------------------------------------------------------------------- [APP]
A ⊢ e1(e2, ... , en) : β, (τ1 = (α2, ... , αn) → α) ∧ (β ⊆ α) ∧ (τ2 ⊆ α2) ∧ ... ∧ (τn ⊆ αn) ∧ C1 ∧ ... ∧ Cn

A ⊢ p : τ, Cp     A ⊢ g : boolean(), Cg
------------------------------------------ [PAT]
A ⊢ p when g : τ, Cp ∧ Cg

A ∪ {v → τv | v ∈ Var(p1)} ⊢ p1 : α1, Cp1,  b1 : β1, Cb1
....
A ⊢ e : τ, Ce    A ∪ {v → τv | v ∈ Var(pn)} ⊢ pn : αn, Cpn,  bn : βn, Cbn
-------------------------------------------------------------------------------------------------------------------- [CASE]

> A ⊢ case e of p1 → b1; ... pn → bn end : β, Ce ∧ (C1 ∨ ... ∨ Cn) where Ci = ((β = βi) ∧ (τi = αi) ∧ Cpi ∧ Cbi)

Reviewer: ```suggestion
A ⊢ case e of p1 → b1; ... pn → bn end : β, Ce ∧ (C1 ∨ ... ∨ Cn) where Ci = ((β = βi) ∧ (τ = αi) ∧ Cpi ∧ Cbi)
```

This is a bug of the paper.

Author: Thanks!

This file shows derivation rules used in fialyzer.

Our derivation rules are almost same as [the original paper](https://it.uu.se/research/group/hipe/papers/succ_types.pdf)'s one, but extended by remote call, local call, list, etc.

```
A : context (mapping of variable to type)
τ : type
x : variable
e : term
fun f/a : local function (f: function name, a: arity)
fun m:f/a : remote function (m: module name, f: function name, a: arity)
```

```
-------------------------- [VAR]
A ∪ {x → τ} ⊢ x : τ, ∅

A ⊢ e1 : τ1, C1 ... en : τn, Cn
------------------------------------------------------ [STRUCT]
A ⊢ {e1, ... , en} : {τ1, ... , τn}, C1 ∧ ... ∧ Cn

A ⊢ e1 : τ1, C1         A ∪ {x → τ1} ⊢ e2 : τ2, C2
---------------------------------------------------- [LET]
A ⊢ let x = e1 in e2 : τ2, C1 ∧ C2

A ∪ {fun xi/ai → τi} ⊢ f1 : τ'1, C1 ... fn : τ'n, Cn        e : τ, C
---------------------------------------------------------------------------------------------- [LETREC]

> A ⊢ letrec x1 = f1, ... , xn = fn in e : τ, C1 ∧ ... Cn ∧ C ∧ (τ'1 = τ1) ∧ ... ∧ (τ'n = τn)

Author: Thanks!

> A ⊢ letrec x1 = f1, ... , xn = fn in e : τ, C1 ∧ ... Cn ∧ C ∧ (τ'1 = τ1) ∧ ... ∧ (τ'n = τn)

Author: fixed https://github.com/dwango/fialyzer/pull/164/commits/db2a446288766391e20f220484303f47e3df8ae9

This file shows derivation rules used in fialyzer.

Our derivation rules are almost same as [the original paper](https://it.uu.se/research/group/hipe/papers/succ_types.pdf)'s one, but extended by remote call, local call, list, etc.

```
A : context (mapping of variable to type)
τ : type
α, β : type variable
x : variable
e : term
fun f/a : local function (f: function name, a: arity)
fun m:f/a : remote function (m: module name, f: function name, a: arity)

> ```

Reviewer 2: it is good if `p : pattern` is written.

Author: Indeed. Thanks!

Author: fixed https://github.com/dwango/fialyzer/pull/164/commits/5bb44eba7ebf07f37788d64d91ec002dcf52f645

```
-------------------------- [VAR]
A ∪ {x → τ} ⊢ x : τ, ∅

A ⊢ e1 : τ1, C1 ... en : τn, Cn
------------------------------------------------------ [STRUCT]
A ⊢ {e1, ... , en} : {τ1, ... , τn}, C1 ∧ ... ∧ Cn

A ⊢ e1 : τ1, C1         A ∪ {x → τ1} ⊢ e2 : τ2, C2
---------------------------------------------------- [LET]
A ⊢ let x = e1 in e2 : τ2, C1 ∧ C2

A ∪ {fun xi/ai → αi} ⊢ f1 : τ1, C1 ... fn : τn, Cn      e : τ, C   where ai = length (args fi)
------------------------------------------------------------------------------------------------- [LETREC]
A ⊢ letrec x1 = f1, ... , xn = fn in e : τ, C1 ∧ ... Cn ∧ C ∧ (τ1 = α1) ∧ ... ∧ (τn = αn)

A ∪ {x1 → α1, ... , xn → αn} ⊢ e : τ, C
---------------------------------------------------------------------- [ABS]
A ⊢ fun(x1, ... , xn) → e : (α1, ... , αn) → τ, C

A ⊢ e : τ, C  e1 : τ1, C1 ... en : τn, Cn
-------------------------------------------------------------------------------------------------------------- [APP]
A ⊢ e(e1, ... , en) : β, (τ = (α1, ... , αn) → α) ∧ (β ⊆ α) ∧ (τ1 ⊆ α1) ∧ ... ∧ (τn ⊆ αn) ∧ C ∧ C1 ∧ ... ∧ Cn

A ⊢ p : τ, Cp     A ⊢ g : boolean(), Cg
------------------------------------------ [PAT]

> A ⊢ p when g : τ, Cp ∧ Cg

Reviewer: In the our implementation of the derivation, a subtyping constraint of the guard expressions is added.
```ocaml
A ⊢ p : τ, Cp     A ⊢ g : τg, Cg
------------------------------------------ [PAT]
A ⊢ p when g : τ, τg ⊆ boolean() ∧ Cp ∧ Cg
```

Author: You are right. Thanks!

Author: fixed https://github.com/dwango/fialyzer/pull/164/commits/c7aeb50aefc05d4bcaeb4cfd69246df0603c9d94

                 A ∪ {v → αv | v ∈ Var(p1)} ⊢ p1 : τp1, Cp1,  b1 : τb1, Cb1
                                           ...
A ⊢ e : τ, Ce    A ∪ {v → αv | v ∈ Var(pn)} ⊢ pn : τpn, Cpn,  bn : τbn, Cbn
-------------------------------------------------------------------------------------------------------------------- [CASE]
A ⊢ case e of p1 → b1; ... pn → bn end : β, Ce ∧ (C1 ∨ ... ∨ Cn) where Ci = ((β = τbi) ∧ (τ = τpi) ∧ Cpi ∧ Cbi)

------------------------------------- [LOCALFUN]
A ∪ {fun f/a → τ} ⊢ fun f/a : τ, ∅

----------------------------------------- if m and f is atom literal, a is non_neg_integer literal [MFA]
A ∪ {fun m:f/a → τ} ⊢ fun m:f/a : τ, ∅

A ⊢ m : τm, Cm   A ⊢ f : τf, Cf   A ⊢ a : τa, Ca
-------------------------------------------------------------------------------------------------- if either m, f, a is not atom literal or non_neg_integer literal [MFAEXPR]

> A ⊢ fun m:f/a : τ, (τ ⊆ any()) ∧ (τm ⊆ atom()) ∧ (τf ⊆ atom()) ∧ (τa ⊆ number()) ∧ Cm ∧ Cf ∧ Ca

Reviewer: I made a issue https://github.com/dwango/fialyzer/issues/172.

> A ⊢ fun m:f/a : τ, (τ ⊆ any()) ∧ (τm ⊆ atom()) ∧ (τf ⊆ atom()) ∧ (τa ⊆ number()) ∧ Cm ∧ Cf ∧ Ca

Author: Oh..I'll remove it. Thanks!

> A ⊢ fun m:f/a : τ, (τ ⊆ any()) ∧ (τm ⊆ atom()) ∧ (τf ⊆ atom()) ∧ (τa ⊆ number()) ∧ Cm ∧ Cf ∧ Ca

Author: I'll fix it. Thanks!

> A ⊢ fun m:f/a : τ, (τ ⊆ any()) ∧ (τm ⊆ atom()) ∧ (τf ⊆ atom()) ∧ (τa ⊆ number()) ∧ Cm ∧ Cf ∧ Ca

Author: fixed https://github.com/dwango/fialyzer/pull/164/commits/52faacc0ed138526ea75579f931a2c658da61043

> A ⊢ fun m:f/a : τ, (τ ⊆ any()) ∧ (τm ⊆ atom()) ∧ (τf ⊆ atom()) ∧ (τa ⊆ number()) ∧ Cm ∧ Cf ∧ Ca

Author: fixed https://github.com/dwango/fialyzer/pull/164/commits/52faacc0ed138526ea75579f931a2c658da61043

> A ⊢ fun m:f/a : τ, (τ ⊆ any()) ∧ (τm ⊆ atom()) ∧ (τf ⊆ atom()) ∧ (τa ⊆ number()) ∧ Cm ∧ Cf ∧ Ca

Author: Oops, We shouldn't care it in this PR, and should care in #172.
I'll revert.

> A ⊢ fun m:f/a : τ, (τ ⊆ any()) ∧ (τm ⊆ atom()) ∧ (τf ⊆ atom()) ∧ (τa ⊆ number()) ∧ Cm ∧ Cf ∧ Ca

Author: https://github.com/dwango/fialyzer/pull/164/commits/af3e29701c3dc8d3c4d09dc0bfa0effb0b02b823

> A ⊢ fun m:f/a : τ, (τ ⊆ any()) ∧ (τm ⊆ atom()) ∧ (τf ⊆ atom()) ∧ (τa ⊆ number()) ∧ Cm ∧ Cf ∧ Ca

Reviewer: There is no need to remove this because our implementation has `τ ⊆ any()` now.
see https://github.com/dwango/fialyzer/blob/f6a430648637e715914095a9c3dea31a42bdda4b/lib/derivation.ml#L114
We needs to fix the implementation. → https://github.com/dwango/fialyzer/issues/172

> A ⊢ fun m:f/a : τ, (τ ⊆ any()) ∧ (τm ⊆ atom()) ∧ (τf ⊆ atom()) ∧ (τa ⊆ number()) ∧ Cm ∧ Cf ∧ Ca

Author: Yes, I noticed that right after I removed. I've already reverted it. Thank you!

> A ⊢ fun m:f/a : τ, (τ ⊆ any()) ∧ (τm ⊆ atom()) ∧ (τf ⊆ atom()) ∧ (τa ⊆ number()) ∧ Cm ∧ Cf ∧ Ca

Author: done by https://github.com/dwango/fialyzer/pull/164/commits/7ce573f05e9ba9162a8ba83f9186a593258ba6b7 (#172 is resolved)

```

> The following derivation rules are the differences from the original paper.

Reviewer: How about it like the following?
```suggestion
The differences from the derivation rules on the original paper are as follows.
```

Author: I'll replace with your suggestion!

Author: fixed https://github.com/dwango/fialyzer/pull/164/commits/c211189783aab91a8fc6d4a4fcb8684837cdad39

                                       ...
A ⊢ e : τ, Ce    An ⊢ pgn : τpgn, Cpgn    An ⊢ bn : τbn, Cbn      where Ai = A ∪ {v → αv | v ∈ Var(pi)}
-------------------------------------------------------------------------------------------------------------------- [CASE]
A ⊢ case e of pg1 → b1; ... pgn → bn end : β, Ce ∧ (C1 ∨ ... ∨ Cn) where Ci = ((β = τbi) ∧ (τ = τpgi) ∧ Cpgi ∧ Cbi)

------------------------------------- [LOCALFUN]
A ∪ {fun f/a → τ} ⊢ fun f/a : τ, ∅

----------------------------------------- if m and f is atom literal, a is non_neg_integer literal [MFA]
A ∪ {fun m:f/a → τ} ⊢ fun m:f/a : τ, ∅

A ⊢ m : τm, Cm   A ⊢ f : τf, Cf   A ⊢ a : τa, Ca
------------------------------------------------------------------------------------ if neither m, f is atom literal nor a is non_neg_integer literal [MFAEXPR]
A ⊢ fun m:f/a : β, (τm ⊆ atom()) ∧ (τf ⊆ atom()) ∧ (τa ⊆ number()) ∧ Cm ∧ Cf ∧ Ca
```

## Differences from the original paper

The differences from the derivation rules on the original paper are as follows.

- α, β, and τ are clearly distinguished. τ is a type, and α, β are type variables.
- LET is fixed: `e2`, not `e`.

> - LETREC is modified: restricted by local functions.

Author: I realized that I misunderstood original LETREC rule...
I misunderstood that variable `f` means an arbitrary expression in original LETREC rule. But actually, `f` means a function.
For this misunderstanding and the current implementation of `derivation.ml` (LETREC is used for only local functions), I noted "restricted by local functions".

So, I'm thinking about reverting LETREC rule. Do you think about it?

> - LETREC is modified: restricted by local functions.

Author: OK. Thank you for your review!

> - LETREC is modified: restricted by local functions.

Author: done https://github.com/dwango/fialyzer/pull/164/commits/53435028f1d009169a9ab22ce409fac5a244a4e9
